### PR TITLE
Issue 373 - Add proper Kinesis output support

### PR DESCRIPTION
**Pull request by trink**

### CODE REVIEW

File: aws/sandboxes/heka/output/kinesis.lua [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+-- Due to the way Kinesis is billed it makes senses to pack multiple items
+-- into a single message to reduce the PUT Payload Unit charges. Warning: if
+-- pack is greater than zero and and pack_partitions is true (multiple batches)
+-- the check pointing system cannot track them individualy i.e. in case of a
+-- failure items could be lost.
+pack_percentage = 0 -- fill percentage of the put request (0 (no packing) - <100).
+         -- The record won't be sent unless the current PUT Payload unit is this
+         -- percentage full or the record has reached its maximum size or the
+         -- timed out is triggered.
```
Comment (ameihm0912) [inline aws/sandboxes/heka/output/kinesis.lua]: time out

File: aws/sandboxes/heka/output/kinesis.lua [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+pack_delimiter = nil    -- Used to add a record delimiter, if necessary
+pack_partitions = false -- create a batch for each partition
```
Comment (ameihm0912) [inline aws/sandboxes/heka/output/kinesis.lua]: should we add a reference to what a partition is here (e.g., create a batch for each partition where data is partitioned based on partitionField)

File: aws/sandboxes/heka/output/kinesis.lua [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+require "aws.kinesis"
+require "string"
+require "table"
+require "xxhash"
+
+local PUT_MAX   = 1024 * 1024
+local PUT_UNIT  = 25 * 1024
+
+local streamName            = read_config("streamName") or error"streamName must be set"
+local partitionField        = read_config("partitionField") or "Uuid"
```
Comment (ameihm0912) [inline aws/sandboxes/heka/output/kinesis.lua]: is Uuid a good default to use if pack_partitions is true but the partition field isn't set? Should we just error out if pack partitions is enabled and the field isn't specified?

File: aws/sandboxes/heka/output/kinesis.lua [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+local encode = require(encoder_module).encode
+if not encode then
+    error(encoder_module .. " does not provide an encode function")
+end
+
+local pack_percentage = read_config("pack_percentage") or 0
+local pack_partitions = read_config("pack_partitions")
+assert(pack_percentage >= 0 and pack_percentage < 100, "0 <= pack_percentage < 100")
+pack_percentage = pack_percentage / 100 * PUT_UNIT
+local pack_delimiter  = read_config("pack_delimiter")
```
Comment (ameihm0912) [inline aws/sandboxes/heka/output/kinesis.lua]: looks like an extra space in here

File: aws/sandboxes/heka/output/kinesis.lua [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+end
+
+if pack_percentage > 0 then
+    local function get_shard_count()
+        local retry = 0
+        repeat
+            local osc, err = producer:open_shard_count(streamName)
+            if osc then return osc end
+            print("open_shard_count", err)
+            retry = retry + 1
```
Comment (ameihm0912) [inline aws/sandboxes/heka/output/kinesis.lua]: is this is a retryable error should we include a delay here?

File: aws/sandboxes/heka/output/kinesis.lua (hunk 3 above)
Comment (trink) [reply aws/sandboxes/heka/output/kinesis.lua]: Sounds good, updating

File: aws/sandboxes/heka/output/kinesis.lua (hunk 5 above)
Comment (trink) [reply aws/sandboxes/heka/output/kinesis.lua]: The throttling is in the underlying API call

**Pull request merged**